lang/cs/Org.Apache.REEF.Network.Examples.Client/ElasticBroadcastClientWithFailEvaluatorBeforeWorkflow.cs

```diff
@@ -0,0 +1,53 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using Org.Apache.REEF.Driver;
+using Org.Apache.REEF.Tang.Interface;
+using Org.Apache.REEF.Tang.Util;
+using Org.Apache.REEF.Utilities.Logging;
+using Org.Apache.REEF.Network.Examples.Elastic;
+
+namespace Org.Apache.REEF.Network.Examples.Client
+{
+    public sealed class ElasticBroadcastClientWithFailEvaluatorBeforeWorkflow : ElasticBroadcastClient
+    {
+        public ElasticBroadcastClientWithFailEvaluatorBeforeWorkflow(bool runOnYarn, int numTasks, int startingPortNo, int portRange)
+            : base (runOnYarn, numTasks, startingPortNo, portRange)
+        {
+        }
+
+        protected override string JobIdentifier
+        {
+            get { return "ElasticBroadcastWithFailure"; }
+        }
+
+        protected override IConfiguration GetDriverConf()
+        {
+            return DriverConfiguration.ConfigurationModule
+                    .Set(DriverConfiguration.OnDriverStarted, GenericType<ElasticBroadcastDriverWithFailEvaluatorBeforeWorkflow>.Class)
+                    .Set(DriverConfiguration.OnEvaluatorAllocated, GenericType<ElasticBroadcastDriverWithFailEvaluatorBeforeWorkflow>.Class)
+                    .Set(DriverConfiguration.OnEvaluatorFailed, GenericType<ElasticBroadcastDriverWithFailEvaluatorBeforeWorkflow>.Class)
+                    .Set(DriverConfiguration.OnContextActive, GenericType<ElasticBroadcastDriverWithFailEvaluatorBeforeWorkflow>.Class)
+                    .Set(DriverConfiguration.OnTaskRunning, GenericType<ElasticBroadcastDriverWithFailEvaluatorBeforeWorkflow>.Class)
+                    .Set(DriverConfiguration.OnTaskCompleted, GenericType<ElasticBroadcastDriverWithFailEvaluatorBeforeWorkflow>.Class)
+                    .Set(DriverConfiguration.OnTaskFailed, GenericType<ElasticBroadcastDriverWithFailEvaluatorBeforeWorkflow>.Class)
+                    .Set(DriverConfiguration.OnTaskMessage, GenericType<ElasticBroadcastDriverWithFailEvaluatorBeforeWorkflow>.Class)
+                    .Set(DriverConfiguration.CustomTraceLevel, Level.Info.ToString())
+                    .Build();
+        }
```

Comment:
In fact, we can make it even shorter, and have a single client for all drivers. If we parameterize `ElasticBroadcastClient` like this:
```C#
public class ElasticBroadcastClient<T> where T : DefaultElasticDriver
{
    public ElasticBroadcastClient(
        bool runOnYarn, int numTasks, int startingPortNo, int portRange, string jobIdentifier)
    {
        IConfiguration driverConfig = TangFactory.GetTang()
            .NewConfigurationBuilder(GetDriverConf())
            // ...



    private static IConfiguration GetDriverConf()
    {
        return DriverConfiguration.ConfigurationModule
            .Set(DriverConfiguration.OnDriverStarted, GenericType<T>.Class)
            // ...
```
then in `Org.Apache.REEF.Network.Examples.Client.Run` switch we can write e.g.
```C#
// ...
if (TestType.ElasticBroadcastWithFailEvaluatorBeforeWorkflow.Match(testToRun))
{
    new ElasticBroadcastClient<ElasticBroadcastDriverWithFailEvaluatorBeforeWorkflow>(
        runOnYarn, 
        numNodes, 
        startPort, 
        portRange,
        "ElasticBroadcast");
```
etc. (and I have a hunch that? in a similar manner, we can have a single generic driver, too) :-)

Reply:
Yeah, it looks doable - we can parameterize, say `ElasticBroadcastDriverWithFailures` as
```C#
public abstract class ElasticBroadcastDriverWithFailures<TMaster, TSlave>
    : DefaultElasticDriver
    where TMaster : DefaultElasticTask
    where TSlave : DefaultElasticTask
{
    // ...
```
so that your `Run` will have invocations like
```C#
new ElasticBroadcastClient<
    ElasticBroadcastDriverWithFailures<BroadcastMasterTask, BroadcastSlaveTask>>(...)
```
(plus now you can make the constructor private and expose only a static method instead)

Also, `ElasticBroadcastDriverWithFailures.MasterTaskConfiguration` can now be just
```C#
private IConfiguration MasterTaskConfiguration(string taskId)
{
    return TangFactory.GetTang().NewConfigurationBuilder(
        Context.GetTaskConfigurationModule(taskId)
            .Set(TaskConfiguration.Task, GenericType<TMaster>.Class)
            .Build())
        .Build();
}
```

Reply:
This is a great suggestion! Fixed.